xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+ *      Copyright (C) 2005-2015 Team XBMC
+ *      http://xbmc.org
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "DVDClock.h"
+#include "DVDDemuxMultiSource.h"
+#include "DVDDemuxUtils.h"
+#include "DVDFactoryDemuxer.h"
+#include "DVDInputStreams/DVDInputStream.h"
+#include "utils/log.h"
+
+
+CDVDDemuxMultiSource::CDVDDemuxMultiSource()
+{
+}
+
+CDVDDemuxMultiSource::~CDVDDemuxMultiSource()
+{
+  Dispose();
+}
+
+void CDVDDemuxMultiSource::Abort()
+{
+  for (auto iter : m_pDemuxers)
+    iter->Abort();
+}
+
+void CDVDDemuxMultiSource::Dispose()
+{
+  while (!m_demuxerQueue.empty())
+  {
+    m_demuxerQueue.pop();
+  }
+
+  m_streamIdToDemuxerMap.clear();
+  m_DemuxerToInputStreamMap.clear();
+  m_pDemuxers.clear();
+  m_currentDemuxer = NULL;
+  m_pInput = NULL;
+
+}
+
+void CDVDDemuxMultiSource::Flush()
+{
+  for (auto iter : m_pDemuxers)
+    iter->Flush();
+}
+
+int CDVDDemuxMultiSource::GetNrOfStreams()
+{
+  int streamsCount = 0;
+  for (auto iter : m_pDemuxers)
+    streamsCount += iter->GetNrOfStreams();
+
+  return streamsCount;
+}
+
+CDemuxStream* CDVDDemuxMultiSource::GetStream(int iStreamId)
```

Comment:
those methods are called quite often. I am wondering if it makes sense to cache streams and update the cache on changes